### ISSUE TITLE
add rule : Control flow statements "if", "for", "while", "switch", "try", ... should not be nested too deeply

### ISSUE THREAD

**Issue by TristanFAURE** [palantir/tslint#4276]
### Rule Suggestion

**Is your rule for a general problem or is it specific to your development style?**
General

**What does your suggested rule do?**
it helps readability of the code

**List several examples where your rule could be used**
Aerospace, Avionic, ... This rules is required for european space industry

**Additional context**
https://rules.sonarsource.com/c/RSPEC-134
https://rules.sonarsource.com/java/RSPEC-134

**Comment by JoshuaKGoldberg** [comment]
@TristanFAURE this is a good idea that might already exist in [`cyclomatic-complexity`](https://palantir.github.io/tslint/rules/cyclomatic-complexity/). Does that work for your use case?

**Comment by TristanFAURE** [comment]
thank you for the link, maybe it could be implemented through the cyclomatic complexity rule but AFAIK cyclomatic computes the differents paths not the number of nested construct (our standards explicitly ask us to have a specific number of maximum levels, 5 for instance) ?

**Comment by JoshuaKGoldberg** [comment]
I'm under the impression it does specify nesting. Do you have any code examples that can't be automatically flagged by `cyclomatic-complexity`? If there are situations where that rule falls short, we could take in a new rule or enhancement to `cyclomatic-complexity`.

**Comment by DamienGarrido** [comment]
The cyclomatic complexity simply counts the branch statements in the control flow until the maximum specified threshold is reached, whereas the nesting depth can be incremented and decremented before the maximum specified threshold is eventually reached.
Thus, as an example, with a configured cyclomatic complexity of 3 the following code won't pass, whereas with a configured nesting depth of 3 this code will obviously pass:
```typescript
function nestingDepthOfOneCyclomaticComplexityOfFour(val: number) {
    if (val === 1) {
        return true;
    }
    if (val === 2) {
        return true;
    }
    if (val === 3) {
        return true;
    }
    return false;
}
```

**Comment by JoshuaKGoldberg** [comment]
Excellent, that clears it up nicely. Thanks! Totally agreed on the value of this proposal. Internally, I'm thinking of it as the _vertical (depth)_ complexity, rather than the _horizontal_ complexity.

Less important: "cyclomatic" refers to the number of different paths through a function. Is there an equivalent word that can be used for this new rule name? Would `nested-complexity` be understandable? Is there precedent here?

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for new rules. See #4534. 😱

If you'd like to see this rule implemented, you have two choices:
* **Recommended:** Use [typescript-eslint](https://typescript-eslint.io) to lint your TypeScript code with ESLint ✅
* **Legacy:** Write it as a [custom TSLint rule](https://palantir.github.io/tslint/develop/custom-rules/) 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
